### PR TITLE
automatically fetch and generate protobuffer definitions

### PR DESCRIPTION
**Pull request by SamuelScheit**
## Changes
- Implemented the `getModules` function, because it isn't available (in none patched discord versions)
- Added a puppeteer-script that opens discord, executes parse.js and writes the protobuffer definitions
- Added javascript protobuffer implementations with types
- Modified .gitignore, .prettierrc.yml, package.json
- Added a github action to automatically update and push the definitions
- Updated the Readme
- Published it as a npm package

@dolfies Please write me your npm username so I can add you as an package owner.

### CODE REVIEW

**Comment by arHSM** [comment]
> Implemented the `getModules` function, because it isn't available (in none patched discord versions)
`getModules` was intentionally not implemented.

> Added a puppeteer-script that opens discord, executes parse.js and writes the protobuffer definitions
The script was supposed to be able to executed by the user in any discord client they wished to do so, limiting it to puppeteer limits where and what client the proto definitions are being extracted from, what if i'm not logged into web version? what if i don't want proto definitions from the stable version?

> Added javascript protobuffer implementations with types
The goal for this repository was just `.proto` definition files and nothing more.

> Added a github action to automatically update and push the definitions
That's unnecessary, Discord-Datamining//Discord-Datamining monitors the main js chunk and they notify for any proto changes, the script can be ran accordingly as changes in the protocol buffers is not frequent.

> Published it as a npm package
If anyone for whatsoever reason wanted to use Discord's protocol definitions they will most probably download the definitions from this repository or another repository providing them and integrate them into their workflow as opposed to downloading an npm package. what if i'm using python?

**Comment by splatterxl** [comment]
> changes in the protocol buffers is not frequent.

Then the cron job should be updated to check maybe every day instead of a very frequent timespan. 

> what if i'm using python?

Then you would integrate the protos themselves into your project instead of complaining that a npm package exists. 

> The goal for this repository was just .proto definition files and nothing more.

I’d like to hear this from the repository maintainer himself, otherwise I believe in the spirit of OSS that contributions be considered. 

——

IMO the output files should be stored on a different branch, but that’s just my style choice to avoid too many GitHub actions commits on master.

**Comment by SamuelScheit** [comment]
@arHSM
> getModules was intentionally not implemented.

I've added an example implementation to be able to use the script without Better Discord. Notice if the getModules function is already available in the global scope it does not overwrite it.

> The script was supposed to be able to executed by the user in any discord client they wished to do so, limiting it to puppeteer limits where and what client the proto definitions are being extracted from, what if i'm not logged into web version? what if i don't want proto definitions from the stable version?

I did not change the script, it still works the same as before.
I've made an additional puppeteer program that loads up discord and executes the script.

> The goal for this repository was just .proto definition files and nothing more.

Then why are their js/py scripts to use the proto definition files to generate js/py code from them?
And even if that would be true, you can extend the goal of the project.

> That's unnecessary, Discord-Datamining//Discord-Datamining monitors the main js chunk and they notify for any proto changes, the script can be ran accordingly as changes in the protocol buffers is not frequent.

Why do it manually, when you can automate it ;)


> If anyone for whatsoever reason wanted to use Discord's protocol definitions they will most probably download the definitions from this repository or another repository providing them and integrate them into their workflow as opposed to downloading an npm package. what if i'm using python?

You can still manually download it, the npm package is just easier to use for js projects.

**Comment by arHSM** [comment]
I'm fine with the proposed changes, was just nitpicking 😅

File: .github/workflows/update.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+name: update definitions
```
Comment (dolfies) [inline .github/workflows/update.yml:1]:
```suggestion
name: Update definitions
```

File: .github/workflows/update.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    permissions:
+      # Give the default GITHUB_TOKEN write permission to commit and push the changed files back to the repository.
+      contents: write
+
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        ref: ${{ github.head_ref }}
+    - uses: actions/setup-node@v3
```
Comment (dolfies) [inline .github/workflows/update.yml:20]:
```suggestion
        ref: ${{ github.head_ref }}

    - uses: actions/setup-node@v3
```

File: .github/workflows/update.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      with:
+        node-version: 18
+    - run: npm install
+    - run: npm run load
+    - run: npm run build
+    - run: npm run test
+
+    - uses: stefanzweifel/git-auto-commit-action@v4
+      with:
+        commit_message: Update protobuffer definitions
```
Comment (dolfies) [inline .github/workflows/update.yml:30]:
```suggestion
        commit_message: Update protobuf definitions
```

**Review by dolfies — CHANGES_REQUESTED** [review]
Thanks for the PR! Just a couple of very small nitpicks.

My npm username is [`@dolfies`](https://www.npmjs.com/~dolfies) btw

**Review by dolfies — APPROVED** [review]

**Comment by SamuelScheit** [comment]
Awesome 👍 I've committed your changes and sent you an invitation on npm

**Comment by dolfies** [comment]
Thanks again!

**Pull request merged**